File: partiql-ast/src/main/java/org/partiql/ast/v1/expr/ExprSessionAttribute.java
package org.partiql.ast.v1.expr;

import lombok.Builder;
import org.jetbrains.annotations.NotNull;
import org.partiql.ast.v1.AstNode;
import org.partiql.ast.v1.AstVisitor;

import java.util.Collection;
import java.util.Collections;

/**
 * TODO docs, equals, hashcode
 */
@Builder(builderClassName = "Builder")
public class ExprSessionAttribute extends Expr {
    @NotNull
    public final SessionAttribute sessionAttribute;

    public ExprSessionAttribute(@NotNull SessionAttribute sessionAttribute) {
        this.sessionAttribute = sessionAttribute;
    }

    @Override
    @NotNull
    public Collection<AstNode> children() {
        return Collections.emptyList();
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitExprSessionAttribute(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/expr/ExprSessionAttribute.kt
package org.partiql.ast.v1.expr

import org.partiql.ast.v1.AstNode
import org.partiql.ast.v1.AstVisitor

/**
 * TODO docs, equals, hashcode
 */
public class ExprSessionAttribute(
    @JvmField
    public var attribute: Attribute,
) : Expr() {
    public override fun children(): Collection<AstNode> = emptyList()

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitExprSessionAttribute(this, ctx)

    /**
     * TODO docs, equals, hashcode
     */
    public enum class Attribute {
        CURRENT_USER,
        CURRENT_DATE,
        OTHER,
    }
}
